proposals/generic-ref-struct.md

```diff
@@ -0,0 +1,424 @@
+
+- `where T : struct, ref struct`: describes a type parameter that can be a `struct` or `ref struct` but not other types like `class` or `interface`.
+- `where T : unmanaged, ref struct`: similar to `struct, ref struct` except that the type also meets the `unmanaged` constraint. This is believed to have limited use though as `ref struct` containing any type of `ref` data will not meet the `unmanaged` contraint and hence are likely better defined as simply `struct`.
+- `where T : notnull, ref struct`: this is the same as the `notnull` constraint but also allows for `ref struct` types to satisfy it.
+- `where T : class, ref struct`: is an error as the `ref struct` portion is meaningless when combined with `class`. 
+- `where T : default, ref struct`: is an error as the declarations are contradictory.
+- `where T : ref struct, BaseType` where `BaseType` denotes a reference type. This is an error because the `ref sruct` portion is meaningless in this context.
+
+The `where T : ref struct` annotation will be encoded by attributing the type parameter with the attribute `System.Runtime.CompilerServices.SupportsByRefLike`. This will not result in a `modreq` on the containing member nor will `[Obsolete]` tricks be added to the containing type. The intent is for this to be a binary compatible change. The explicit intent is for us to allow types in libraries to relax their type parameters to accept `ref struct` without breaking consumers.
+
+### Supporting Span<T>
+A key motivation of this proposal is to allow for instantiations like `Span<Span<T>>`. Effectively moving the `Span<T>` type to have the `where T : ref struct` annotation. This is a problem because a number of the APIs on `Span<T>` violate the rules of `where T : ref struct` as `T[]` is a common type in the APIs. 
+
+```c# 
+// Existing Span<T> definition
+public readonly ref struct Span<T>
+{
+    public Span(T[]? array) { ... }
+
+    public Span(T[]? array, int index, int length) { ... }
+
+    public static implicit operator Span<T>(T[]? array) { ... }
+}
+```
+
+If the `Span<T>` type was being designed from scratch in the presence of `where T : ref struct` it's possible different choices would've been made here. Possibly leveraging `static` factories and pushing implicit conversions into the compiler. Unfortunately this is not being designed from scratch and in order to meet the desire to have `Span<Span<T>>` these existing APIs need to be rationalized.
+
+The compiler will introduce the type `System.Runtime.CompilerServices.SupportsOnlyNonByRefLikeAttribute` that can be applied to methods and operators. When applied to a member it has the effect of removing `where T : ref struct` from the constraints on all type parameters on the *immediate* containing type. All other constraints remain in place.
```

Comment:
Is there any room to consider a more general feature, where users can further constrain specific methods on types, so that a method only applies to the subset of generic instantiations that fulfill the increased constraints?

Currently it's common to do this by implementing extension methods, but this doesn't work for cases where you need access to private data, or for static methods.

If this was allowed you could add a `not ref struct` constraint which all type parameters have by default, and write:

```csharp
public Span(T[]? array) where T : not ref struct { ... }
```

I don't think this is incredibly useful, but I prefer general elegant solutions over specific hacky ones, and given the plethora of extension methods which do this there definitely will be plenty of use cases which would have preferred an instance method, but were forced to use an extension method.

Reply:
This is effectively inching into generic specialization (though really, it's anti-specialization?), and I wonder what the thoughts from the runtime team on generalized specialization are.

Reply:
> I don't think this is incredibly useful, but I prefer general elegant solutions over specific hacky ones, and given the plethora of extension methods which do this there definitely will be plenty of use cases which would have preferred an instance method, but were forced to use an extension method.

Why do you feel this is a hacky solution vs a general one? This proposal is effectively relaxing a restriction that prevented an entire class of types from being used in generic arguments. I agree the syntax can seem off at first, which is why there are a few listed alternatives, but conceptually this is a general feature. 

> If this was allowed you could add a not ref struct constraint which all type parameters have by default, and write:

The default though is effectively not `ref struct` though aWhy do we need a new syntax here?

Reply:
@jaredpar I was referring to `System.Runtime.CompilerServices.SupportsOnlyNonByRefLikeAttribute` specifically

Reply:
I think my main concern with the name is that it doesn't specify that it is changing the meaning of type parameters. It makes it sound a little like "also, you can't use spans and stuff in here"

Reply:
Okay. Yes that is definitely an icky part of this design. 

My original idea was to just special case the methods in .NET that needed this change. Essentially the way we special case some behaviors of arrays. After some discussions I was talked into it being a more general feature. As I wrote it up though I wasn't happy with the result. Been waffling on just making it back to a special case.